lang/interpolate/parse.rl

```
@@ -0,0 +1,180 @@
+		# non-consecutive dots.
+		var_name
+			= ( [a-z] ([a-z0-9_]
+			  | ('.' | '_') [a-z0-9_])*
+			  )
+			>start
+			@{ v.Name = data[idx:fpc+1] }
+			;
+
+		# var is a reference to a variable.
+		var = '${' var_name '}'
+			;
+
+		#escape_sequence = ( '\\a'
+		#		  | '\\b'
+		#		  | '\\f'
+		#		  | '\\n'
+		#		  | '\\r'
+		#		  | '\\t'
+		#		  | '\\v'
+		#		  )
+		#	>start
+		#	@{ l = Literal{Value: "XXX" } }
+		#	;
+
+		# Anything followed by a '\' is matched and used appropriately.
+		#escaped_lit = '\\' any @{ l = Literal{Value: data[fpc:fpc+1]} };
+		#//l = Literal{Value: "Z("+data[fpc:fpc+1]+"):["+ data[fpc-1:fpc+1] +"]" }
+
+		# works!
+		#escaped_lit = '\\' any
+		#@{
+		#	switch s := data[fpc:fpc+1]; s {
+		#	case "a":
+		#		x = "\a"
+		#	case "b":
+		#		x = "\b"
+		#	//case "e":
+		#	//	x = "\e" // non-standard
+		#	case "f":
+		#		x = "\f"
+		#	case "n":
+		#		x = "\n"
+		#	case "r":
+		#		x = "\r"
+		#	case "t":
+		#		x = "\t"
+		#	case "v":
+		#		x = "\v"
+		#	case "\\":
+		#		x = "\\"
+		#	case "\"":
+		#		x = "\""
+		#	case "$":
+		#		x = "$"
+		#	//case "0":
+		#	//	x = "\x00"
+		#	default:
+		#		//panic(fmt.Sprintf("unhandled escaped: %s", s))
+		#		x = s // XXX: should we error cleanly instead?
+		#	}
+		#	l = Literal{Value: x}
+		#};
+
+		# Any special escape characters are matched here.
+		escaped_lit = '\\' ( 'a' | 'b' | 'f' | 'n' | 'r' | 't' | 'v' | '\\' | '"' | '$' )
+		@{
+			switch s := data[fpc:fpc+1]; s {
+			case "a":
+				x = "\a"
+			case "b":
+				x = "\b"
+			//case "e":
+			//	x = "\e" // non-standard
+			case "f":
+				x = "\f"
+			case "n":
+				x = "\n"
+			case "r":
+				x = "\r"
+			case "t":
+				x = "\t"
+			case "v":
+				x = "\v"
+			case "\\":
+				x = "\\"
+			case "\"":
+				x = "\""
+			case "$":
+				x = "$"
+			//case "0":
+			//	x = "\x00"
+			default:
+				//panic(fmt.Sprintf("unhandled escaped: %s", s))
+				x = s // XXX: should we error cleanly instead?
```

It makes sense to raise an `error` of `invalid escape sequence '$rune'` or similar. That's what most compilers would do IMO

---

Mostly because a user might be expecting a particular behaviour and silent failures are a nightmare to debug